src/empathy-accounts.1
.TH EMPATHY-ACCOUNTS "1" "February 2010" "Telepathy project" "User Commands"
.SH NAME
empathy-accounts \- Accounts configuration dialog for Empathy instant messaging client
.SH SYNOPSIS
empathy-accounts
.RI [ options ]
.SH DESCRIPTION
Accounts configuration dialog for Empathy instant messaging client
.PP
This utility allows you to configure your instant messaging accounts for Empathy.
.SH OPTIONS
.TP
.BR \-h ,\  \-\-hidden
Don't display any dialogs; do any work (eg, importing) and exit
.TP
.BR \-n ,\  \-\-if\-needed
Don't display any dialogs if there are any non-salut accounts
.TP
.BI \-s\  account\-id
.TQ
.BI \-\-select\-account= account\-id
Initially select given account (eg, gabble/jabber/foo_40example_2eorg0)
.SH AUTHOR
This manual page was written by
.MT bigon@debian.org
Laurent Bigonville
.ME ,
for the Debian GNU/Linux system (but may be used by others).
.SH SEE ALSO
.BR empathy (1),
.UR http://telepathy.freedesktop.org/
The Telepathy project
.UE ,
.UR http://live.gnome.org/Empathy
The Empathy project
.UE .


NEWS
NEW in 2.31.5.1 (13/07/2010)
===============

A small release fixing detection of libnm-glib which was accidentally
broken in 2.31.5.

Bugs fixed:
 - Fixed #624202, empathy requires libnm_glib instead of libnm-glib (Laurent Bigonville)

Translations:
 - Updated gl Translation (Fran Diéguez)
 - Updated he Translation (Yaron Shahrabani)

NEW in 2.31.5 (12/07/2010)
=============

The "Please popup chats in my face!" release.

Empathy finally gained an option to automatically present incoming events
rather than making the status icon blink.

The notifications bubbles now contain more buttons (if your notify server
supports actions) allowing you to quickly accept or decline incoming events.

The audio/video client now lives in its own process: 'empathy-av'.

Empathy can now build with GTK+ 3. If GTK+ 3 is present during configure it
will be used and Empathy will requier updated versions of other dependencies:
  - libcanberra-gtk3 ≥ 0.25
  - libnotify ≥ 0.5.1
  - libunique ≥ 2.90.1
  - WebKitGTK+ ≥ 1.3.2 (optional)
  - nautilus-sendto ≥ 2.90.0 (optional)
If GTK+ 3 is not detected or if --enable-gtk3=no is passed to configure, then
GTK+ 2 is used and the other dependencies are unchanged.


Bugs fixed:
 - Fixed #585914, Notification icon must be clicked before user can see new messages (Guillaume Desmottes)
 - Fixed #597124, Should make better use of libnotify actions (Guillaume Desmottes)
 - Fixed #599161, Audio/Video client should be a separated process (Guillaume Desmottes)
 - Fixed #617405, Empathy's short (menu) description should focus on its use, not the technology (Guillaume Desmottes)
 - Fixed #621753, Should be able to build with GTK+3 (Guillaume Desmottes)
 - Fixed #622729, duplicated messages when opening chat (Guillaume Desmottes)
 - Fixed #623054, Should re-add data/org.gnome.Empathy.gschema.xml.in to POTFILES.in (Danielle Madeley)
 - Fixed #623112, Bip backlog are not displayed any more (Guillaume Desmottes)
 - Fixed #623117, Debug window should look for TP clients (Guillaume Desmottes)
 - Fixed #623358, Use TpSimpleHandler to handle text channels (Guillaume Desmottes)
 - Fixed #623396, Fixed some leaks (Guillaume Desmottes)
 - Fixed #623679, /join broken (Guillaume Desmottes)
 - Fixed #623756, empathy-CRITICAL **: empathy_chatroom_is_always_urgent: assertion `EMPATHY_IS_CHATROOM (chatroom)' failed (Sjoerd Simons)
 - Fixed #623817, Sync man pages with Debian (Laurent Bigonville, Guillaume Desmottes)
 - Fixed #623914, since tp-logger history is shown as normal messages in adium themes (Guillaume Desmottes)
 - Fixed #623937, empathy -h does not start hidden (Guillaume Desmottes)
 - Fixed #623972, Toggling [ ] Enabled in the accounts dialog with the keyboard doesn't work (Will Thompson)
 - Fixed #624054, Fix a memory leak in EmpathyAvatar (Philip Withnall)

Translations:
 - Updated es Translation (Jorge González)
 - Updated et Translation (Ivar Smolin)
 - Updated gl Translation (Fran Diéguez)
 - Updated he Translation (Yaron Shahrabani)
 - Updated nb Translation (Kjartan Maraas)
 - Updated sl Translation (Matej Urbančič)

NEW in 2.31.4 (28/06/2010)
=============

A lot of new things in this release!

Empathy now has an awesome contact live search feature.
See http://kaserf.blogspot.com/2010/06/empathy-live-search.html for details.

Empathy doesn't log messages any more but relies on a new component,
the telepathy-logger, to do so. As a result the feature allowing you to tag
contacts as favourites is now built-in.

The contact list and status icon are now a proper Telepathy Approver. That
means external applications can now approve channels as well and Empathy will
handle them nicely.

This release makes a few steps forward to the bright GNOME 3 future by removing
deprecated GTK+ symbols and switching from GConf to GSettings.

Dependencies:

* telepathy-logger ≥ 0.1.2 (0.1.3 recommended)
* telepathy-glib ≥ 0.11.7
* libunique ≥ 0.1.2
* GTK+ ≥ 2.21.2

Bugs fixed:
 - Fixed #586476, Use accessor functions instead direct access (Danielle Madeley)
 - Fixed #588922, Empathy should use ContactInfo (Xavier Claessens)
 - Fixed #589225, We should check if the contact supports vnc stream tube for Share my desktop (Guillaume Desmottes)
 - Fixed #599158, Contact list and status icon should be an Approver (Guillaume Desmottes)
 - Fixed #608166, Add contacts live search bar (Felix Kaser, Xavier Claessens)
 - Fixed #610956, Should depend on telepathy-logger (Guillaume Desmottes)
 - Fixed #616362, Port Empathy to GSettings (Danielle Madeley)
 - Fixed #619873, search-as-you-type should only match beginnings of words (Xavier Claessens)
 - Fixed #620903, some messages get dropped when using adium themes (Guillaume Desmottes)
 - Fixed #621023, empathy does not set geoclue requirements before asking for address/position (Laurent Bigonville)
 - Fixed #621089, The live search don't always have focus in main window (Xavier Claessens)
 - Fixed #621409, empathy-2.30.1.1: tests fail (Guillaume Desmottes)
 - Fixed #621545, Stable empathy passes GTK_DISABLE_SINGLE_INCLUDES and doesn't require libunique that is fine with that (Xavier Claessens)
 - Fixed #621703, Here come a couple of UI tweaks (Will Thompson)
 - Fixed #621789, crash when closing chat with still-open notification bubble (Guillaume Desmottes)
 - Fixed #621797, Doesn't display outgoing message (Guillaume Desmottes)
 - Fixed #621894, Should port to new tp-logger API (Guillaume Desmottes)
 - Fixed #621998, Should stop using ShareWithTube() (Guillaume Desmottes)
 - Fixed #622201, Can't connect chatroom if space after adress (Guillaume Desmottes)
 - Fixed #622266, Should decide which default font to use for chat view (Xavier Claessens)
 - Fixed #622273, Crash when closing the log viewer while searching (Guillaume Desmottes)
 - Fixed #622368, Audio/Video: Use TpSimpleHandler (Guillaume Desmottes)
 - Fixed #622370, Should bold dates having logs (Guillaume Desmottes)
 - Fixed #622381, Debug dialog: display Logger logs (Guillaume Desmottes)

src/empathy-accounts.c
/*
 * Copyright (C) 2005-2007 Imendio AB
 * Copyright (C) 2007-2010 Collabora Ltd.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA  02110-1301  USA
 *
 * Authors: Martyn Russell <martyn@imendio.com>
 *          Xavier Claessens <xclaesse@gmail.com>
 *          Cosimo Cecchi <cosimo.cecchi@collabora.co.uk>
 *          Jonathan Tellier <jonathan.tellier@gmail.com>
 *          Travis Reitter <travis.reitter@collabora.co.uk>
 */

#include <config.h>

#include <string.h>
#include <stdlib.h>

#include <gtk/gtk.h>
#include <glib/gi18n.h>
#include <unique/unique.h>

#include <telepathy-glib/account-manager.h>
#include <telepathy-glib/defs.h>
#include <telepathy-glib/util.h>

#include <libempathy/empathy-utils.h>
#include <libempathy/empathy-connection-managers.h>
#include <libempathy-gtk/empathy-ui-utils.h>

#include "empathy-accounts.h"
#include "empathy-accounts-common.h"
#include "empathy-accounts-dialog.h"
#include "empathy-account-assistant.h"
#include "empathy-auto-salut-account-helper.h"

#define DEBUG_FLAG EMPATHY_DEBUG_ACCOUNT
#include <libempathy/empathy-debug.h>

#define EMPATHY_ACCOUNTS_DBUS_NAME "org.gnome.EmpathyAccounts"

static gboolean only_if_needed = FALSE;
static gboolean hidden = FALSE;
static gchar *selected_account_name = NULL;

static void
account_prepare_cb (GObject *source_object,
    GAsyncResult *result,
    gpointer user_data)
{
  TpAccountManager *manager = TP_ACCOUNT_MANAGER (user_data);
  TpAccount *account = TP_ACCOUNT (source_object);
  GError *error = NULL;

  if (!tp_account_prepare_finish (account, result, &error))
    {
      DEBUG ("Failed to prepare account: %s", error->message);
      g_error_free (error);

      account = NULL;
    }

  empathy_accounts_show_accounts_ui (manager, account,
      G_CALLBACK (gtk_main_quit));
}

static void
maybe_show_accounts_ui (TpAccountManager *manager)
{
  if (hidden ||
      (only_if_needed && empathy_accounts_has_non_salut_accounts (manager)))
    gtk_main_quit ();
  else
    empathy_accounts_show_accounts_ui (manager, NULL, gtk_main_quit);
}

static void
account_manager_ready_for_accounts_cb (GObject *source_object,
    GAsyncResult *result,
    gpointer user_data)
{
  TpAccountManager *manager = TP_ACCOUNT_MANAGER (source_object);
  const gchar *account_id = (const gchar*) user_data;
  GError *error = NULL;

  if (!tp_account_manager_prepare_finish (manager, result, &error))
    {
      DEBUG ("Failed to prepare account manager: %s", error->message);
      g_clear_error (&error);
      return;
    }

  if (account_id != NULL)
    {
      gchar *account_path;
      TpAccount *account = NULL;
      TpDBusDaemon *bus;

      /* create and prep the corresponding TpAccount so it's fully ready by the
       * time we try to select it in the accounts dialog */
      account_path = g_strdup_printf ("%s%s", TP_ACCOUNT_OBJECT_PATH_BASE,
          account_id);
      bus = tp_dbus_daemon_dup (NULL);
      if ((account = tp_account_new (bus, account_path, &error)))
        {
          tp_account_prepare_async (account, NULL, account_prepare_cb, manager);
          return;
        }
      else
        {
          DEBUG ("Failed to find account with path %s: %s", account_path,
              error->message);
          g_clear_error (&error);
        }

      g_object_unref (bus);
      g_free (account_path);
    }
  else
    {
      maybe_show_accounts_ui (manager);
    }
}

static UniqueResponse
unique_app_message_cb (UniqueApp *unique_app,
    gint command,
    UniqueMessageData *data,
    guint timestamp,
    gpointer user_data)
{
  DEBUG ("Other instance launched, presenting the main window. "
      "Command=%d, timestamp %u", command, timestamp);

  if (command == UNIQUE_ACTIVATE)
    {
      TpAccountManager *account_manager;

      account_manager = tp_account_manager_dup ();

      empathy_accounts_show_accounts_ui (account_manager, NULL,
          G_CALLBACK (gtk_main_quit));

      g_object_unref (account_manager);
    }
  else
    {
      g_warning (G_STRLOC "unhandled unique app command %d", command);

      return UNIQUE_RESPONSE_PASSTHROUGH;
    }

  return UNIQUE_RESPONSE_OK;
}

#define COMMAND_ACCOUNTS_DIALOG 1

int
main (int argc, char *argv[])
{
  TpAccountManager *account_manager;
  GError *error = NULL;
  UniqueApp *unique_app;

  GOptionContext *optcontext;
  GOptionEntry options[] = {
      { "hidden", 'h',
        0, G_OPTION_ARG_NONE, &hidden,
        N_("Don't display any dialogs; do any work (eg, importing) and exit"),
        NULL },
      { "if-needed", 'n',
        0, G_OPTION_ARG_NONE, &only_if_needed,
        N_("Don't display any dialogs if there are any non-salut accounts"),
        NULL },
      { "select-account", 's',
        G_OPTION_FLAG_IN_MAIN, G_OPTION_ARG_STRING, &selected_account_name,
        N_("Initially select given account (eg, "
            "gabble/jabber/foo_40example_2eorg0)"),
        N_("<account-id>") },

      { NULL }
  };

  g_thread_init (NULL);
  empathy_init ();

  optcontext = g_option_context_new (N_("- Empathy Accounts"));
  g_option_context_add_group (optcontext, gtk_get_option_group (TRUE));
  g_option_context_add_main_entries (optcontext, options, GETTEXT_PACKAGE);

  if (!g_option_context_parse (optcontext, &argc, &argv, &error))
    {
      g_print ("%s\nRun '%s --help' to see a full list of available command line options.\n",
          error->message, argv[0]);
      g_warning ("Error in empathy init: %s", error->message);
      return EXIT_FAILURE;
    }

  g_option_context_free (optcontext);

  empathy_gtk_init ();

  g_set_application_name (_("Empathy Accounts"));

  gtk_window_set_default_icon_name ("empathy");
  textdomain (GETTEXT_PACKAGE);

  unique_app = unique_app_new (EMPATHY_ACCOUNTS_DBUS_NAME, NULL);

  if (unique_app_is_running (unique_app))
    {
      unique_app_send_message (unique_app, UNIQUE_ACTIVATE, NULL);

      g_object_unref (unique_app);
      return EXIT_SUCCESS;
    }

  account_manager = tp_account_manager_dup ();

  tp_account_manager_prepare_async (account_manager, NULL,
    account_manager_ready_for_accounts_cb, selected_account_name);

  g_signal_connect (unique_app, "message-received",
      G_CALLBACK (unique_app_message_cb), NULL);

  gtk_main ();

  g_object_unref (account_manager);
  g_object_unref (unique_app);

  return EXIT_SUCCESS;
}
